File: src/main/java/datastructures/stack/java/CustomStack.java
package datastructures.stack.java;

import java.util.Arrays;

public class CustomStack<E> {

    private int size;
    private int count = 0;
    private Object[] dataList;

    public CustomStack(int size) {
        this.size = size;
        dataList = new Object[size];
    }

    public E push(E e) {
        if (count == dataList.length) throw new StackOverflowError();
        dataList[count++] = e;
        return e;
    }

    public synchronized E peek() {
        if (count == 0) throw new IllegalStateException("Stack is empty");
        return (E) dataList[count - 1];
    }

    public synchronized E pop() {
        E element = peek();
        count--;
        return element;
    }

    public void displayStack() {
        System.out.println(Arrays.toString(getDataList()));
    }

    public int size() {
        return count;
    }

    public Object[] getDataList() {
        return Arrays.copyOfRange(dataList, 0, count);
    }
}

File: src/main/java/datastructures/stack/kotlin/CustomStack.kt
package datastructures.stack.kotlin

import java.util.*

class CustomStack<E>(var size: Int) {
    var count: Int = 0
        get() = field
    var array: Array<Any?> = arrayOfNulls(size)

    open fun push(element: E): E {
        if (count == size) throw StackOverflowError("Stack overflow error")
        array[count++] = element
        return element;
    }

    open fun peek(): Any? {
        if (count == 0) throw  IllegalStateException("Empty Stack")
        return array[count - 1];
    }

    open fun pop(): Any? {
        var element: Any? = peek()
        count--;
        return element;
    }

    fun displayStack() {
        println(Arrays.toString(getDataList()))
    }

    fun getDataList(): Array<Any?>? {
        return array.copyOfRange(0, count)
    }

    fun isEmpty() : Boolean { return count == 0}

}